lib/logsource.c

```diff
@@ -174,6 +174,95 @@ log_source_flow_control_suspend(LogSource *self)
   window_size_counter_suspend(&self->window_size);
 }
 
+void
+log_source_enable_dynamic_window(LogSource *self, DynamicWindowCounter *window_ctr)
+{
+  if (self->pos_tracked && window_ctr)
+    {
+      msg_warning("WARNING: dynamic window size control works only with non-position tracking sources"
+                  "Falling back to static window size (dynamic won't be used).",
+                  log_pipe_location_tag(&self->super));
+      return;
+    }
+
+  dynamic_window_set_counter(&self->dynamic_window, window_ctr);
+}
+
+void
+log_source_dynamic_window_update_statistics(LogSource *self)
+{
+  dynamic_window_stat_update(&self->dynamic_window, window_size_counter_get(&self->window_size, NULL));
+  msg_trace("Updating dynamic window statistic", evt_tag_int("avg window size",
+                                                             dynamic_window_stat_get_avg(&self->dynamic_window)));
+}
+
+static inline void
+_decrease_window(LogSource *self)
+{
+  gsize new_full_window_size = MAX(self->options->init_window_size, self->full_window_size / 2);
+  gsize subtrahend = self->full_window_size - new_full_window_size;
+
+  gsize empty_window = window_size_counter_get(&self->window_size, NULL);
+  if (empty_window <= subtrahend)
+    {
+      subtrahend = empty_window == 0 ? 0 : empty_window - 1;
+      new_full_window_size = self->full_window_size - subtrahend;
+    }
+
+  window_size_counter_sub(&self->window_size, subtrahend, NULL);
+
+  msg_trace("Decreasing dynamic window", evt_tag_int("previous_window", self->full_window_size),
+            evt_tag_int("new_window", new_full_window_size), log_pipe_location_tag(&self->super));
+
+  self->full_window_size = new_full_window_size;
+  dynamic_window_release(&self->dynamic_window, subtrahend);
+}
+
+static inline void
+_increase_window(LogSource *self)
+{
+  gsize offered_dynamic = dynamic_window_request(&self->dynamic_window, self->full_window_size);
+
+  msg_trace("Increasing dynamic window", evt_tag_int("previous_window", self->full_window_size),
+            evt_tag_int("new_window", self->full_window_size + offered_dynamic), log_pipe_location_tag(&self->super));
+
+  self->full_window_size += offered_dynamic;
+
+  gsize old_window_size = window_size_counter_add(&self->window_size, offered_dynamic, NULL);
+  if (old_window_size == 0 && offered_dynamic != 0)
+    log_source_wakeup(self);
+}
+
+static void
+_release_dynamic_window(LogSource *self)
+{
+  g_assert(self->ack_tracker == NULL);
+
+  gsize dynamic_part = self->full_window_size - self->options->init_window_size;
+  msg_trace("Releasing dynamic part of the window", evt_tag_int("dynamic_window_to_be_released", dynamic_part),
+            log_pipe_location_tag(&self->super));
+  self->full_window_size -= dynamic_part;
```

Comment:
could you make the log tag names consistent? so in all invocation, we have the same names for the same stuff.

sometimes it's previous_window/new_window, now it's dynamic_window_to_be_released?

Also, if I may ask, the evt_tag() functions should be on their own lines, I find them much easier to read that way. Thanks. This last one is not a blocker at all, simply my preference.